core/src/main/java/com/linecorp/armeria/server/misc/Inet4AddressBlock.java

```
@@ -0,0 +1,114 @@
+        final int mask = 1 << 32 - maskBits;
+        // e.g. (-mask) is 0xFFFFFF00 if mask is 256.
+        lowerBound = ipv4AddressToInt(baseAddress.getAddress()) & -mask;
+        // e.g. (mask-1) is 0x000000FF if mask is 256.
+        upperBound = lowerBound + mask - 1;
+    }
+
+    @Override
+    public boolean test(InetAddress address) {
+        requireNonNull(address, "address");
+        if (maskBits == 0) {
+            return true;
+        }
+        final int addr = inetAddressToInt(address);
+        return addr >= lowerBound && addr <= upperBound;
+    }
+
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                          .add("baseAddress", baseAddress)
+                          .add("maskBits", maskBits)
+                          .add("lowerBound", "0x" + Integer.toHexString(lowerBound))
+                          .add("upperBound", "0x" + Integer.toHexString(upperBound))
+                          .toString();
+    }
+
+    /**
+     * Returns the integer representation of the specified {@link InetAddress}.
+     */
+    private static int inetAddressToInt(InetAddress inetAddress) {
+        final byte[] address;
+        if (inetAddress instanceof Inet6Address) {
+            address = ipv6ToIpv4Address((Inet6Address) inetAddress);
+        } else {
+            address = inetAddress.getAddress();
+        }
+        return ipv4AddressToInt(address);
+    }
+
+    /**
+     * Returns IPv4 byte representation of the specified {@link Inet6Address}.
+     *
+     * @throws IllegalArgumentException if the IPv6 cannot be mapped to IPv4
+     */
+    private static byte[] ipv6ToIpv4Address(Inet6Address address) {
+        final byte[] addr = address.getAddress();
+        for (int i = 0; i < 9; i++) {
+            if (addr[i] != 0) {
+                throw new IllegalArgumentException("This IPv6 address cannot be used in IPv4 context");
```

This will `test()` raise an exception rather than returning `false`.